### PR TITLE
Add Circuit.with_noise

### LINKED ISSUE #1610: Apply noise model to circuit
I expected to be able to take a `NoiseModel` and apply it to a circuit.

```noise = cirq.ConstantQubitNoiseModel(cirq.depolarize(0.2))```
Then I'd like to be able to
```circuit.add_noise(noise)```
or 
```noise.apply_noise(circuit)```

### PR DESCRIPTION
**Pull request by cduck**
Fixes #1610.

### CODE REVIEW

File: cirq/circuits/circuit_test.py [hunk 1]
```diff
@@ -3238,3 +3238,42 @@ def test_json_dict():
...
+        cirq.Moment([
+            cirq.X(q0),
+        ]),
+        cirq.Moment([
+            cirq.H(q0),
+        ]),
+    ])
+    c_noisy = c.with_noise(Noise())
+    assert str(c_noisy) == str(c_expected)
+    assert c_noisy == c_expected
```
Comment (dstrain115) [inline cirq/circuits/circuit_test.py]: Why are we testing both string equality and equality?

Comment (Strilanc) [reply cirq/circuits/circuit_test.py]: Probably leftover debugging code.

**Comment by mpharrigan** [comment]
Should noisy simulators be modified to use this functionality?

**Comment by cduck** [comment]
@mpharrigan Ideally, yes.  There are two challenges that need to be resolved to do that:
- The intermediate results simulator needs to know what came from the original moments.
- What do you do if the user specifies a qubit order and the noise model substitutes the qubits (#2103)?

File: cirq/circuits/circuit_test.py (hunk 1 above)
Comment (cduck) [reply cirq/circuits/circuit_test.py]: Fixed.

**Comment by cduck** [comment]
@googlebot Committed from my computer instead of the web editor.

**Comment by Strilanc** [comment]
> Should noisy simulators be modified to use this functionality?

No, because the noisy simulators sometimes have to preserve information about moment structure in a way that this method does not.

**Pull request merged**